src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonValueOfT.cs

```diff
@@ -0,0 +1,371 @@
+            {
+                return ConvertJsonElement<T>();
+            }
+
+            // Currently we do not support other conversions.
+            // Generics (and also boxing) do not support standard cast operators say from 'long' to 'int',
+            //  so attempting to cast here would throw InvalidCastException.
+            throw new InvalidOperationException(SR.Format(SR.NodeUnableToConvert, _value!.GetType(), typeof(T)));
+        }
+
+        public override bool TryGetValue<[DynamicallyAccessedMembers(MembersAccessedOnRead)] T>([NotNullWhen(true)] out T value)
+        {
+            // If no conversion is needed, just return the raw value.
+            if (_value is T returnValue)
+            {
+                value = returnValue;
+                return true;
+            }
+
+            if (_value is JsonElement jsonElement)
+            {
+                return TryConvertJsonElement<T>(out value);
+            }
+
+            // Currently we do not support other conversions.
+            // Generics (and also boxing) do not support standard cast operators say from 'long' to 'int',
+            //  so attempting to cast here would throw InvalidCastException.
+            value = default!;
+            return false;
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2072:ystem.Text.Json.Node.JsonValue<TValue>.WriteTo(Utf8JsonWriter,JsonSerializerOptions): 'inputType' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicFields', 'DynamicallyAccessedMemberTypes.PublicProperties' in call to 'System.Text.Json.JsonSerializer.Serialize(Utf8JsonWriter,Object,Type,JsonSerializerOptions)'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.",
+            Justification = "The 'inputType' parameter if obtained by calling System.Object.GetType().")]
+        public override void WriteTo(Utf8JsonWriter writer, JsonSerializerOptions? options = null)
+        {
+            if (writer == null)
+            {
+                throw new ArgumentNullException(nameof(writer));
+            }
+
+            if (_value is JsonElement jsonElement)
+            {
+                jsonElement.WriteTo(writer);
+            }
+            else
+            {
+                JsonSerializer.Serialize(writer, _value, _value!.GetType(), options);
+            }
+        }
+
+        internal TypeToConvert ConvertJsonElement<TypeToConvert>()
+        {
+            JsonElement element = (JsonElement)(object)_value!;
+            Type returnType = typeof(TypeToConvert);
+            Type? underlyingType = Nullable.GetUnderlyingType(returnType);
+            returnType = underlyingType ?? returnType;
+
+            switch (element.ValueKind)
+            {
+                case JsonValueKind.Number:
+                    if (returnType == typeof(int))
```

Comment:
I believe some of those checks can be eliminated by the JIT if you replace `returnType` with `typeof(TypeToConvert)`.

e.g. https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACAygCzACcAHAGTACMA6AJQFcA7YASygFMBuddOAZmyJsAYWwBvdNinZJ0gNoApFsADi7RuyIsAxgApgATxLsA9gDNdLZgEprAGmxLV6zTv1HTFgDYnGAc1sHJzUNLT1DY3NdABMTekovdkDHZRDXcI8oiHwTImBbAF1ZKX5sK2BsAFkAHgAVAD5da2LxFukys2x3SIta62wAXgHsCM9LG2a0dvaJKen5uAB2bCRuOfmAXzbpbakWTu6xvsHh0aiff1td1vX56SXsBDW77C3b9uv9rrPe/qGRzIWWLxRJXd7SWYvdoPPjPO5vF7XB4AFjhUjeGyAA==

Reply:
Thanks that is a big difference in IL.

Since the previous code normalized the Type above (converting nullable to non-nullable) I needed to add an additional `if` check for each type, however the result is still smaller and faster. I didn't attempt to further optimize it by bifurcating the nullable vs. non-nullable logic.